### PR TITLE
Fix version ranges for non-semver

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
This is an attempt at trying to understand why OTP 24 is not available yet. Opening as a draft as most likely this investigation will result is a closed-unmerged branch.

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
Hm...

* `semver.validRange('24')` returns `'>=24.0.0 <25.0.0-0'`, which is seemingly Okay (`24` is the input spec),
* `semver.maxSatisfying(['24.0'], '>=24.0.0 <25.0.0-0')` returns `null`, which is not usable (`24.0` is the value found in the Hex.pm listing).

[Some might say](https://jubianchi.github.io/semver-check/#/24/24.0) `semver` is wrong, here.

Thoughts on how to solve this?

**Comment by paulo-ferraz-oliveira** [comment]
Lets try with `24.0`, instead of `24`: 51e8d2f.

File: .github/workflows/test.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/test.yml:102]: This appears to not be the best option. Waiting for input.

**Comment by paulo-ferraz-oliveira** [comment]
Moving from Draft, for visibility. Not merge-ready yet.

**Comment by ericmj** [comment]
OTP 24 is not available because Bob has not built it yet. There seems to be a regression in OTP 24 that causes the compilation to fail.

**Comment by ericmj** [comment]
> OTP 24 is not available because Bob has not built it yet. There seems to be a regression in OTP 24 that causes the compilation to fail.

It seems it only fails for Docker images, so the issue is unrelated.

**Comment by paulo-ferraz-oliveira** [comment]
@ericmj, I've tracked it down to this being a `semver` (`npm` package) issue. We either:
* force `24` to be `24.0` (e.g. in a `hacks` module), or
* we create a more generic rule (e.g. `x` becomes `x.0`), or
* we fix `semver`, or
* we find an alternative to `semver`.

I'd prefer to not fiddle with our code, since its assumptions are correct, and either find an alternative to `semver` or open an issue to fix it. I'll start be opening the issue, but accept thoughts 😄

**Comment by paulo-ferraz-oliveira** [comment]
Bug opened: https://github.com/npm/node-semver/issues/382.

**Comment by ericmj** [comment]
I don't believe it's a bug in the semver package but rather that OTP versions are not semver. `24.0` is not valid semver so it's understandable that `semver.maxSatisfying` does not accept it.

We could try to convert `24.0 => 24.0.0` so that the value is accepted.

Personally I don't think we should support version ranges because CI should be reproducible and automatically updating versions are not reproducible but I guess we need to maintain the backwards compatibility. Did this work before the `setup-beam` rewrite?

**Comment by paulo-ferraz-oliveira** [comment]
> I don't believe it's a bug in the semver package but rather that OTP versions are not semver. 24.0 is not valid semver so it's understandable that semver.maxSatisfying does not accept it.

Understandable, yes; that it won't/shouldn't accept it... I'm still waiting for a reply to the issue.

> Personally I don't think we should support version ranges because CI should be reproducible 

Agree almost completely; this is from the "`setup-elixir`" times. We offer this flexibility for people who don't care about repeatability (I guess!), also, and because otherwise this action could just be a couple of `bash` scripts.

> and automatically updating versions are not reproducible

That's the main issue I have with the ranges (non-reproducibility). At the same time, some will use branches in their dependencies, which also break this promise, but on our end, not supporting ranges could be an extra safety net.

> but I guess we need to maintain the backwards compatibility

We don't _need_ to. As per semver we just go 2.x (and add some kind of release note) and that's it 😄 

> Did this work before the setup-beam rewrite?

Most surely not, as per https://github.com/actions/setup-elixir/blob/main/src/setup-elixir.js#L104, which was left untouched, algo-wise.

**Comment by paulo-ferraz-oliveira** [comment]
I find that `semver.minVersion(spec).version` for `spec=24` returns `24.0.0`, as well as for `spec=24.0`, so I'm going with that, for now.

Mind you I changed the `24.0` in the tests back to `24` to test this particular change.

**Comment by paulo-ferraz-oliveira** [comment]
I've squash-rebased all of this. Waiting for approval and merging...

File: src/setup-beam.js
Comment (paulo-ferraz-oliveira) [inline src/setup-beam.js:188]: @starbelly, @ericmj, @wojtekmach, could I ask your 👀 on this so we can merge it and ship 1.7.1? I'd want to pull this to a couple of repo.s, already. Thanks much.

If you have a better, even less verbose for example, "fix", do tell. I'd like to potentially consider dropping the ver-ranges, but not in this pull request.

**Review by starbelly — APPROVED** [review]

**Comment by paulo-ferraz-oliveira** [comment]
I pushed some more tests for the newly introduced code based on no-longer maintained versions (so it won't easily break in a near future).

**Pull request merged**
